server/schedulers/balance_leader.go

```diff
@@ -146,63 +205,108 @@ func (l *balanceLeaderScheduler) Schedule(cluster schedule.Cluster) []*operator.
-				schedulerCounter.WithLabelValues(l.GetName(), "total").Inc()
-				if ops := l.transferLeaderOut(plan); len(ops) > 0 {
-					l.retryQuota.ResetLimit(plan.source)
-					ops[0].Counters = append(ops[0].Counters, l.counter.WithLabelValues("transfer-out", plan.SourceMetricLabel()))
-					return ops
-				}
-			}
-			l.Attenuate(plan.source)
-			log.Debug("no operator created for selected stores", zap.String("scheduler", l.GetName()), zap.Uint64("source", plan.SourceStoreID()))
+	greaterOption := func(stores []*core.StoreInfo, plan *balancePlan) func(int, int) bool {
+		return func(i, j int) bool {
+			iOp := plan.GetOpInfluence(stores[i].GetID())
+			jOp := plan.GetOpInfluence(stores[j].GetID())
+			return stores[i].LeaderScore(plan.kind.Policy, iOp) >
+				stores[j].LeaderScore(plan.kind.Policy, jOp)
+		}
+	}
+	lessOption := func(stores []*core.StoreInfo, plan *balancePlan) func(int, int) bool {
+		return func(i, j int) bool {
+			iOp := plan.GetOpInfluence(stores[i].GetID())
+			jOp := plan.GetOpInfluence(stores[j].GetID())
+			return stores[i].LeaderScore(plan.kind.Policy, iOp) <
+				stores[j].LeaderScore(plan.kind.Policy, jOp)
 		}
-		if i < len(targets) {
-			plan.source, plan.target = nil, targets[i]
-			retryLimit := l.retryQuota.GetLimit(plan.target)
-			log.Debug("store leader score", zap.String("scheduler", l.GetName()), zap.Uint64("target-store", plan.TargetStoreID()))
-			l.counter.WithLabelValues("low-score", plan.TargetMetricLabel()).Inc()
-			for j := 0; j < retryLimit; j++ {
-				schedulerCounter.WithLabelValues(l.GetName(), "total").Inc()
-				if ops := l.transferLeaderIn(plan); len(ops) > 0 {
-					l.retryQuota.ResetLimit(plan.target)
-					ops[0].Counters = append(ops[0].Counters, l.counter.WithLabelValues("transfer-in", plan.TargetMetricLabel()))
-					return ops
+	}
+	sourceCandidate := &candidateStores{
+		stores:        filter.SelectSourceStores(stores, l.filters, cluster.GetOpts()),
+		desc:          "source",
+		actionDesc:    "transfer-out",
+		compareOption: greaterOption,
+		setPlanStore: func(bp *balancePlan, si *core.StoreInfo) {
+			plan.source, plan.target = si, nil
+		},
+		opCreator:     l.transferLeaderOut,
+		schedulerName: l.GetName(),
+		plan:          plan,
+	}
+	sourceCandidate.initSort()
+	targetCandidate := &candidateStores{
+		stores:        filter.SelectTargetStores(stores, l.filters, cluster.GetOpts()),
+		desc:          "target",
+		actionDesc:    "transfer-in",
+		compareOption: lessOption,
+		setPlanStore: func(bp *balancePlan, si *core.StoreInfo) {
+			plan.source, plan.target = nil, si
+		},
+		opCreator:     l.transferLeaderIn,
+		schedulerName: l.GetName(),
+		plan:          plan,
+	}
+	targetCandidate.initSort()
+	usedRegions := make(map[uint64]struct{})
+	candidates := []*candidateStores{sourceCandidate, targetCandidate}
+
+	result := make([]*operator.Operator, 0, l.conf.Batch)
+	for hasStore := true; hasStore; {
+		hasStore = false
+		for _, queue := range candidates {
```

Comment:
I think this loop changed the behavior than before:
```
before:
 - transfer out from hightest store
 - transfer into lowest  store
 -  transfer out from second highest
 ....
 ```
```
after:
 - transfer out from hightest store
 - transfer out from second highest
 ...
 - transfer into lowest  store
```

the new behavior may cause some redundant operators? because we `choose store-> random pick region -> transfer to follower`, the follower‘s store may above the average score.